fusion module
=============

This module uses Arriba to call fusion genes, based on the BAM file produced in the snv-indels modules.

Tools
-----
This module uses the bam file from `STAR <https://github.com/alexdobin/STAR>`_ to
call fusion events with `Arriba <https://github.com/suhrig/arriba/wiki/01-Home>`_.

The fusion events are filtered based on the ``blacklist`` from Arriba itself. Only fusions where at least one of the involved genes is in ``report_genes`` will be included in the final output.

For each fusion event that remains after filtering, we also generate a figure using the ``draw_fusions.R`` script provided by Arriba.

Input
-----
The input for this module is a single bam file, generated by STAR per sample, specified in a PEP configuration file, as is shown below.

.. csv-table:: Example input for the expression module
   :delim: ,
   :file: ../../test/pep/chrM-bam.csv

Output
------
The output of this module are a JSON file with an overview of the most important results, as well as a number of other output files:
- The final Arriba output file, after filtering.
- One figure per fusion event

Configuration
-------------
You can automatically generate a configuration for the fusion module using the ``utilities/create-config.py`` script.

Example
^^^^^^^
.. literalinclude:: ../../test/data/config/fusion.json
   :language: json

Configuration options
^^^^^^^^^^^^^^^^^^^^^
.. list-table:: Configuration options
  :widths: 30 80 15
  :header-rows: 1

  * - Option
    - Description
    - Required
  * - genome_fasta
    - Reference genome, in FASTA format
    - yes
  * - gtf
    - GTF file with transcript information
    - yes
  * - blacklist
    - File of blacklisted fusion events
    - yes
  * - known_fusions
    - A file of known fusion events
    - yes
  * - report_genes
    - Only report fusions involving genes specified in this file
    - yes
  * - cytobands
    - A file with cytoband information for visualization
    - yes
  * - protein_domains
    - A file with protein domains
    - yes
